npms/skale-observer/observer.mjs

```diff
@@ -0,0 +1,1078 @@
+    owaspUtils.ensureObserverOptionsInitialized( opts );
+    if( ! opts.imaState )
+        throw new Error( "Cannot load S-Chain description in observer, no imaState is provided" );
+    if( opts && opts.details ) {
+        if( log.verboseGet() >= log.verboseReversed().trace ) {
+            opts.details.write( cc.debug( "Loading S-Chain " ) + cc.notice( "#" ) +
+                cc.info( idxSChain + 1 ) + cc.debug( " of " ) + cc.info( cntSChains ) +
+                cc.debug( "..." ) + "\n" );
+        }
+    }
+    hash = hash ||
+        await opts.imaState.joSChainsInternal.callStatic.schainsAtSystem(
+            idxSChain, { from: addressFrom } );
+    if( opts && opts.details ) {
+        if( log.verboseGet() >= log.verboseReversed().trace )
+            opts.details.write( cc.debug( "    Hash " ) + cc.attention( hash ) + "\n" );
+    }
+    if( opts && opts.bStopNeeded )
+        return null;
+    let joData =
+        await opts.imaState.joSChainsInternal.callStatic.schains(
+            hash, { from: addressFrom } );
+    joData = owaspUtils.cloneObjectByRootKeys( joData );
+    const joSChain = { "data": joData };
+    removeSChainDescDataNumKeys( joSChain.data, addressFrom );
+    if( opts && opts.bStopNeeded )
+        return null;
+    await loadSChainParts( joSChain, addressFrom, opts );
+    if( opts && opts.details ) {
+        if( log.verboseGet() >= log.verboseReversed().trace ) {
+            opts.details.write( cc.debug( "    Desc " ) + cc.j( joSChain.data ) + "\n" );
+            opts.details.write( cc.success( "Done" ) + "\n" );
+        }
+    }
+    joSChain.isConnected = false;
+    return joSChain;
+}
+
+export async function loadSChains( addressFrom, opts ) {
+    owaspUtils.ensureObserverOptionsInitialized( opts );
+    if( ! opts.imaState )
+        throw new Error( "Cannot load S-Chains in observer, no imaState is provided" );
+    const cntSChains = await getSChainsCount( addressFrom, opts );
+    if( opts && opts.details ) {
+        if( log.verboseGet() >= log.verboseReversed().trace ) {
+            opts.details.write( cc.debug( "Have " ) + cc.info( cntSChains ) +
+                cc.debug( " S-Chain(s) to load..." ) + "\n" );
+        }
+    }
+    const arrSChains = [];
+    for( let idxSChain = 0; idxSChain < cntSChains; ++ idxSChain ) {
+        if( opts && opts.bStopNeeded )
+            break;
+        const joSChain = await loadSChain( addressFrom, idxSChain, null, cntSChains, opts );
+        if( ! joSChain )
+            break;
+        arrSChains.push( joSChain );
+    }
+    if( opts && opts.details ) {
+        if( log.verboseGet() >= log.verboseReversed().trace ) {
+            opts.details.write( cc.success( "All " ) + cc.info( cntSChains ) +
+                cc.debug( " S-Chain(s) loaded:" ) + cc.j( arrSChains ) + "\n" );
+        }
+    }
+    return arrSChains;
+}
+
+export async function loadCachedSChainsSimplified( addressFrom, opts ) {
```

Comment:
Here as well you can use multicall and do 2 calls instead of 21

Reply:
Done